timeboost-sequencer/src/decrypt.rs

```diff
@@ -1,37 +1,453 @@
+            if let Some(status) = self.incls.get_mut(&r) {
+                match status {
+                    Status::Encrypted(incl) => {
+                        // reassemble inclusion list for round r
+                        assemble_incl(r, incl, dec, &mut self.modified)?;
+                    }
+                    Status::Decrypted(_) => {}
+                }
+            };
+
+            // inclusion lists are processed in order; return only if r is decrypted.
+            if let Some(entry) = self.incls.first_entry() {
+                match entry.get() {
+                    Status::Decrypted(incl) => return Ok(incl.clone()),
+                    Status::Encrypted(_) => {
+                        debug!(
+                            "received decrypted txns for r={} but the next round is r={}",
+                            r,
+                            entry.key()
+                        );
+                    }
+                }
+            }
         }
-        Ok(self.forever.clone().await)
+        Err(DecryptError::Shutdown)
+    }
+}
+
+fn assemble_incl(
+    r: RoundNumber,
+    incl: &mut InclusionList,
+    dec: Vec<DecryptedItem>,
+    modified: &mut BTreeMap<RoundNumber, (Vec<usize>, Vec<Transaction>)>,
+) -> Result<(), DecryptError> {
+    let (modified_ptxs, modified_txs) = modified
+        .remove(&r)
+        .expect("encrypted incl => modified entry");
+    let (mut ptxs, mut txs) = incl.to_owned().into_transactions();
+    if modified_ptxs.len() + modified_txs.len() != dec.len() {
+        return Err(DecryptError::State);
+    }
+    modified_ptxs
+        .into_iter()
+        .enumerate()
+        .map(|(i, m)| {
+            if let Some(ptx) = ptxs.get_mut(m) {
+                *ptx = PriorityBundle::new_with_hash(
+                    ptx.epoch(),
+                    ptx.seqno(),
+                    dec.get(i).ok_or(DecryptError::State)?.0.clone(),
+                    ptx.kid(),
+                );
+                Ok::<(), DecryptError>(())
+            } else {
+                Err(DecryptError::State)
+            }
+        })
+        .collect::<Result<Vec<_>, DecryptError>>()?;
+
+    modified_txs
+        .into_iter()
+        .enumerate()
+        .map(|(i, tx)| {
+            if let Some(tx) = txs.take(&tx) {
+                let data = dec
+                    .get(i + ptxs.len())
+                    .ok_or(DecryptError::InvalidMessage)?
+                    .0
+                    .clone();
+                let dec_tx = Transaction::new(tx.nonce(), tx.to(), data, tx.kid());
+                txs.insert(dec_tx);
+                Ok(())
+            } else {
+                Err(DecryptError::State)
+            }
+        })
+        .collect::<Result<Vec<_>, DecryptError>>()?;
+
+    Ok(())
+}
+
+impl Drop for Decrypter {
+    fn drop(&mut self) {
+        self.jh.abort()
+    }
+}
+
+type Incubator = Arc<Mutex<BTreeMap<DecShareKey, Vec<DecShare>>>>;
```

Comment:
```suggestion
type Incubator = BTreeMap<DecShareKey, Vec<DecShare>>;
```
AFAICS `Incubator` is only used in `Worker` and does not require a Mutex, right?

Reply:
thanks for spotting this. think this was a mutex in an early draft and forgot to change it. fixed.